core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala

```diff
@@ -366,6 +385,309 @@ class GroupCoordinatorTest extends JUnitSuite {
+
+    EasyMock.reset(replicaManager)
+    val followerResponseFuture = sendJoinGroup(groupId, JoinGroupRequest.UNKNOWN_MEMBER_ID, protocolType, protocolSuperset, followerInstanceId)
+    // The goal for two timer advance is to let first group initial join complete and set newMemberAdded flag to false. Next advance is
+    // to trigger the rebalance as needed for follower delayed join. One large time advance won't help because we could only populate one
+    // delayed join from purgatory and the new delayed op is created at that time and never be triggered.
+    timer.advanceClock(GroupInitialRebalanceDelay + 1)
+    timer.advanceClock(DefaultRebalanceTimeout + 1)
+    val newGeneration = 1
+
+    val leaderJoinGroupResult = await(leaderResponseFuture, 1)
+    assertEquals(Errors.NONE, leaderJoinGroupResult.error)
+    assertEquals(newGeneration, leaderJoinGroupResult.generationId)
+
+    val followerJoinGroupResult = await(followerResponseFuture, 1)
+    assertEquals(Errors.NONE, followerJoinGroupResult.error)
+    assertEquals(newGeneration, followerJoinGroupResult.generationId)
+
+    val leaderId = leaderJoinGroupResult.memberId
+
+    EasyMock.reset(replicaManager)
+    val syncGroupFuture = sendSyncGroupLeader(groupId, leaderJoinGroupResult.generationId, leaderId, Map(leaderId -> Array[Byte]()))
+    timer.advanceClock(1)
+    val syncGroupResult = Await.result(syncGroupFuture, Duration(1, TimeUnit.MILLISECONDS))
+    assertEquals(Errors.NONE, syncGroupResult._2)
+    assertTrue(getGroup(groupId).is(Stable))
+    (leaderId, followerJoinGroupResult.memberId, newGeneration)
+  }
+
+  private def checkJoinGroupResult(joinGroupResult: JoinGroupResult,
+                                   expectedError: Errors,
+                                   expectedGeneration: Int,
+                                   expectedLeaderId: String,
+                                   expectedMemberId: String,
+                                   expectedMemberSize: Int,
+                                   groupId: String,
+                                   expectedGroupState: GroupState) {
+    assertEquals(Errors.NONE, joinGroupResult.error)
+    assertEquals(expectedGeneration, joinGroupResult.generationId)
+    assertEquals(expectedLeaderId, joinGroupResult.leaderId)
+    assertEquals(expectedMemberId, joinGroupResult.memberId)
+    assertEquals(expectedMemberSize, joinGroupResult.members.size)
+    assertTrue(getGroup(groupId).is(expectedGroupState))
+  }
+
+  @Test
+  def staticMemberRejoinAsDynamicMember() {
+    var joinGroupResult = staticJoinGroup(groupId, JoinGroupRequest.UNKNOWN_MEMBER_ID, groupInstanceId, protocolType, protocols)
+    assertEquals(Errors.NONE, joinGroupResult.error)
+    EasyMock.reset(replicaManager)
+
+    val assignedMemberId = joinGroupResult.memberId
+    joinGroupResult = dynamicJoinGroup(groupId, assignedMemberId, protocolType, protocols)
```

Comment:
Not sure of the implications here, should we somehow work on removing the static member when the same `memberId` re-joins as a dynamic member? I guess it might not hurt, it will eventually get removed when the member leaves or its heartbeat fails, but we will continue to lock that `group.instance.id` with the `member.id` so new joins from static members with that group.instaince.id won't work. In other words, if a consumer becomes a static member and later re-joins as a dynamic, that `group.instance.id` is still taken. Am I correct?

That might not be unwanted behavior though

Reply:
Hey Stanis, the condition you proposed here is not possible within the current setup. Membership type transformation has to go through service restart, which will inevitably reset the member.id. So there is no way we see a dynamic member joining with its member.id points to a known static member. 
However this is a vaild concern, which I think by enforcing an assertion would be safer!